xscen/aggregate.py

```diff
@@ -467,3 +491,282 @@ def spatial_mean(
+
+    yearly_diff = right_column - mean_base  # difference from reference
+    last_year = right_column.iloc[-1].name  # last available year
+
+    # get the star and end date of the window when the warming level is first reached
+    for year in np.arange(1850, 2101):
+        start_yr = int(year - (window / 2 - 1))
+        end_yr = int(year + window / 2)
+        if end_yr > last_year:  # put nan if end_yr is not available in the data
+            start_yr = np.nan
+            end_yr = np.nan
+            break
+        rolling_mean = yearly_diff.loc[start_yr:end_yr].mean().values
+        if rolling_mean >= wl:
+            break
+        else:  # if it never goes above wl
+            start_yr = np.nan
+            end_yr = np.nan
+
+    if np.isnan(start_yr):
+        logger.info(
+            f"Global warming level of +{str(wl)}C is never reached for {id_ds}."
+        )
+        return None
+
+    # cut the window selected above
+    ds_wl = ds.sel(time=slice(str(start_yr), str(end_yr))).load()
+
+    # compute indicators
+    ind_dict = compute_indicators(ds=ds_wl, indicators=indicators)
+
+    # Compute the window-year mean
+    ds_merge = xr.Dataset()
+    concats = []
+    for freq, ds_ind in ind_dict.items():
+        ds_mean = climatological_mean(
+            ds_ind.dropna(
+                dim="time", how="all"
+            ),  # to drop first and last nan DJF, if not we start a year too early
+            window=window,
+            min_periods=min_window,
+            to_level=to_level.format(wl=wl),
+        )
+
+        if "AS" not in freq:  # there is only one time, can't infer_freq
+            # name new_dim
+            if "QS" in freq:
+                new_dim = "season"
+            elif "MS" in freq:
+                new_dim = "month"
+            else:
+                new_dim = "period"
+                warnings.warn(
+                    f"Frequency {freq} is not supported. Setting name of the new dimension to 'period'."
+                )
+            ds_mean = unstack_dates(
+                ds_mean,
+                # seasons = seasons,
+                new_dim=new_dim,
+            )
+            horizon = ds_mean.horizon.values[0, 0]
+            ds_mean = (
+                ds_mean.drop_vars("horizon")
+                .assign_coords(horizon=("time", [horizon]))
+                .swap_dims({"time": "horizon"})
+                .drop_vars("time")
+            )
+
+        else:
+            ds_mean = ds_mean.swap_dims({"time": "horizon"}).drop_vars("time")
+
+        ds_mean["horizon"] = [f"+{str(wl)}C"]
+        concats.append(ds_mean)
+
+        # put all indicators in one dataset
+        for var in ds_mean.data_vars:
+            ds_merge[var] = ds_mean[var]
+        ds_merge.attrs.update(ds_mean.attrs)
+
+    ds_merge.attrs["cat:xrfreq"] = "fx"
+    ds_merge.attrs["cat:frequency"] = "fx"
+    return ds_merge
+
+
+# TODO: include chris' data
+
+
+@parse_config
+def produce_horizon(
```

Comment:
Is the only difference between this and previous function that this does the classical horizons (e.g. 2041-2070), while the other produces the warming levels? If so, I think this goes with my previous comment. This feels more like a mini-workflow, rather than a function.

There's a place for that, but not in `aggregate`... Maybe in `templates`? We should discuss in person about this, it's gonna be easier than in here.

Reply:
yes, let's discuss in person

Reply:
I think produce_horizon fits in aggregate, prepare_warminglevel could maybe go in extract ?

Reply:
Yeah, sounds good to me